lib/logger: Implement suppression (fixes #6132)

calmh opened the pull request:
With this change, certain identical log messages are filtered out,
reducing log volume when there are large numbers of repeats.

The mechanism is as described in #6132.

Oh and I fixed the historical licensing of the logging package.

imsodin commented on lib/logger/suppression.go:
```diff
@@ -0,0 +1,171 @@
+
+	if l.suppressSameAsLast(hash) {
+		// Message is the same as the last message
+		return nil
```
Does this address any concrete case? I couldn't think of any. And in general I'd rather keep repeat messages if suppression is not active/volume is low. I believe that we should not hide them, but correct them where they occur when someone notices (e.g. the repeat warnings in quic recently).

TLDR: I wouldn't suppress anything unless active/high volume.

imsodin commented on lib/logger/logger.go:
```diff
@@ -33,6 +36,12 @@ const (
 
+const (
+	suppressionBuckets  = 5
+	supressionThreshold = 100
```
double-p-typo

imsodin commented on lib/logger/suppression.go:
```diff
@@ -0,0 +1,171 @@
+	}
+
+	l.buckets[0].count++
+	return l.next.Output(level+1, message)
```
Why +1?

calmh commented:
I removed the last message suppression as it's not strictly necessary.

Using utils.AsService would shave maybe a line of code, but isn't possible because it would cause an import loop (util uses sync, sync uses logger)...

I'm guessing both you misunderstood the `level` parameter to be about the log level, which it's not. In this case it's the number of levels to go up to find the source line of the log caller. I think that makes the +1 and the zero make sense, holler if you think it doesn't.

calmh commented:
Are you all certain we need an off switch for this? I mean, it's our decision as an application what to output as log messages, and we have several other places where we for example just output a warning once even though the thing might happen/fail repeatedly.

As for this being attacking it from the wrong end (@imsodin) -- what do you suggest? The main scenario when this would kick in, except for bugs, is when there's a large amount of repeated pull failures. Perhaps we should keep state for each such failure and just log it the first time?

imsodin commented:
> I'm guessing both you misunderstood the level parameter to be about the log level, [...]

Exactly, thanks for the explanation.

> Are you all certain we need an off switch for this?

Not certain no, quite the opposite: I believe this will make debugging easier by removing noise, but I am not certain that I won't ever be stock enough while debugging to suspect that there's some signal in that suppressed noise. So feel free to discount my view as the FUD it is xD

As to the wrong end: My main sentiment is that this is too much complexity and resources for a limited problem. Otherwise what you said: filter/optimize logging where we know there's lots of logging or fix the bug if logging gets unreasonably verbose. And for the latter case have an unsophisticated fuse that curbs logging when unreasonable (if the first 100MB of log don't contain enough info to track the bug, the next 10GB probably wont either).

For pulling I propose https://github.com/syncthing/syncthing/pull/6149

AudriusButkevicius commented:
I'd still like a off switch for this, even if its an env var.

I don't mind this change per-se, I think it will prevent a few ugly cases we had recently.

calmh commented:
I agree filtering log commands in the puller (and so on) is better, and we should do that as you've proposed. I still think this has a place as a "fuse", and given that I think we can't do it much less complex or it will either be useless or eat valuable log data...

imsodin commented:
What about just counting while not active, and starting to hash only once the activity threshold has been exceeded?

calmh commented:
Good suggestion!

calmh commented:
So, I think this works, but do we need it? We have fixed the puller output, which is the large logging culprit in the normal case. We have added rotation, which avoids murdering Windows folks' systems by default. We can still get stuck in logging loops due to bugs, but the rotation will act as a fuse for that and for people with systemd/syslog setups that's unfortunate but their log systems can do deduplication as required... I think we're good without this for now.
